src/main/java/com/netflix/conductor/kafka/eventqueue/KafkaConsumeObservableQueue.java

```diff
@@ -0,0 +1,316 @@
+            records.forEach(record -> {
+                Map<String, String> headers = new HashMap<>();
+                record.headers().forEach(header -> headers.put(header.key(),
+                        new String(header.value(), StandardCharsets.UTF_8)));
+                if (this.filteringHeader != null && !this.filteringValue.equals(headers.get(this.filteringHeader))
+                        || this.filterByPartitionKey && !this.filteringValue.equals(record.key())) {
+                    putDiscardedMsg(record.topic(), record.partition(), record.offset());
+                    return;
+                }
+                putAcceptedMsg(record.topic(), record.partition());
+                String headersJson = null;
+                try {
+                    headersJson = objectMapper.writeValueAsString(headers);
+                } catch (JsonProcessingException e) {
+                    logger.error("message headers build error.", e);
+                }
+                logger.debug("Consumer Record: key: {}, value: {}, partition: {}, offset: {}",
+                        record.key(), record.value(), record.partition(), record.offset());
+                String id = record.key() + MSG_KEY_SEPARATOR + record.topic() + MSG_KEY_SEPARATOR +
+                        record.partition() + MSG_KEY_SEPARATOR + record.offset();
+                String messagePayload = String.format("{\"eventData\":%s,\"eventHeaders\":%s}",
+                        record.value(), headersJson);
+
+                Message message = new Message(id, messagePayload, "");
+                messages.add(message);
+            });
+            commitAsyncOffsets();
+            Monitors.recordEventQueueMessagesProcessed(QUEUE_TYPE, this.queueURI, messages.size());
+        } catch (KafkaException e) {
+            logger.error("kafka consumer message polling failed.", e);
+            Monitors.recordObservableQMessageReceivedErrors(QUEUE_TYPE);
+        }
+        return messages;
+    }
+
+    private void putAcceptedMsg(String topic, int partition) {
+        TopicPartition topicPartition = new TopicPartition(topic, partition);
+        commitOffsetMap.get().remove(topicPartition);
+    }
+
+    private void putDiscardedMsg(String topic, int partition, long offset) {
+        TopicPartition topicPartition = new TopicPartition(topic, partition);
+        commitOffsetMap.get().compute(topicPartition, (k, v) -> v == null ? offset : Math.max(v, offset));
+    }
+
+    private void commitAsyncOffsets() {
+        Map<TopicPartition, OffsetAndMetadata> currentOffsets;
+        for (Map.Entry<TopicPartition, Long> offsetEntry : commitOffsetMap.get().entrySet()) {
+            currentOffsets = new HashMap<>();
+            currentOffsets.put(offsetEntry.getKey(),
+                    new OffsetAndMetadata(offsetEntry.getValue() + 1, "no metadata"));
+            try {
+                consumer.commitSync(currentOffsets);
```

Comment:
Why we need to call commit sync for each partition. Can we call it for all partitions all together?